File: results/0.25/ICMP/icmp-claude-3-5-sonnet-20241022.c
#include <hammer/hammer.h>
#include <stdio.h>
#include <stdlib.h>

// ICMP Type values
#define ECHO_REPLY 0
#define DEST_UNREACHABLE 3
#define SOURCE_QUENCH 4
#define REDIRECT 5
#define ECHO_REQUEST 8
#define TIME_EXCEEDED 11
#define PARAM_PROBLEM 12
#define TIMESTAMP 13
#define TIMESTAMP_REPLY 14
#define INFO_REQUEST 15
#define INFO_REPLY 16

typedef struct {
    uint8_t type;
    uint8_t code;
    uint16_t checksum;
    union {
        struct {
            uint16_t identifier;
            uint16_t sequence;
        } echo;
        struct {
            uint8_t pointer;
            uint8_t unused[3];
        } param_problem;
        uint32_t gateway_addr;
        uint32_t unused;
    } rest_of_header;
    uint8_t* data;
} ICMP_Packet;

static HParsedToken* act_echo(const HParseResult* p, void* user_data) {
    return h_make_seqn(p);
}

static HParsedToken* act_param_problem(const HParseResult* p, void* user_data) {
    return h_make_seqn(p);
}

static HParsedToken* act_redirect(const HParseResult* p, void* user_data) {
    return h_make_seqn(p);
}

static HParsedToken* act_default(const HParseResult* p, void* user_data) {
    return h_make_seqn(p);
}

static HParser* create_icmp_parser(void) {
    // Basic field parsers
    HParser* type = h_uint8();
    HParser* code = h_uint8();
    HParser* checksum = h_uint16();
    
    // Rest of header parsers
    HParser* identifier = h_uint16();
    HParser* sequence = h_uint16();
    HParser* pointer = h_uint8();
    HParser* unused_24bits = h_repeat_n(h_uint8(), 3);
    HParser* gateway_addr = h_uint32();
    HParser* unused_32bits = h_uint32();

    // Data section parser (variable length)
    HParser* data = h_many(h_uint8());

    // Echo/Echo Reply format
    HParser* echo_format = h_sequence(identifier, sequence, data, NULL);

    // Parameter Problem format
    HParser* param_problem_format = h_sequence(pointer, unused_24bits, data, NULL);

    // Redirect format
    HParser* redirect_format = h_sequence(gateway_addr, data, NULL);

    // Default format (with unused 32 bits)
    HParser* default_format = h_sequence(unused_32bits, data, NULL);

    // Complete ICMP parser with proper action functions
    return h_sequence(type, code, checksum,
        h_choice(h_action(h_sequence(h_ch(ECHO_REPLY), echo_format, NULL), act_echo, NULL),
                h_action(h_sequence(h_ch(ECHO_REQUEST), echo_format, NULL), act_echo, NULL),
                h_action(h_sequence(h_ch(PARAM_PROBLEM), param_problem_format, NULL), act_param_problem, NULL),
                h_action(h_sequence(h_ch(REDIRECT), redirect_format, NULL), act_redirect, NULL),
                h_action(default_format, act_default, NULL),
                NULL),
        NULL);
}

int main(int argc, char** argv) {
    if (argc != 2) {
        fprintf(stderr, "Usage: %s <input_file>\n", argv[0]);
        return 1;
    }

    FILE* file = fopen(argv[1], "rb");
    if (!file) {
        perror("Failed to open file");
        return 1;
    }

    fseek(file, 0, SEEK_END);
    size_t size = ftell(file);
    fseek(file, 0, SEEK_SET);

    uint8_t* input = malloc(size);
    if (!input) {
        perror("Failed to allocate memory");
        fclose(file);
        return 1;
    }

    if (fread(input, 1, size, file) != size) {
        perror("Failed to read file");
        free(input);
        fclose(file);
        return 1;
    }

    HParser* icmp_parser = create_icmp_parser();
    HParseResult* result = h_parse(icmp_parser, input, size);

    if (result) {
        printf("Successfully parsed ICMP packet\n");
        h_parse_result_free(result);
    } else {
        printf("Failed to parse ICMP packet\n");
    }

    free(input);
    fclose(file);
    return 0;
}

File: results/0.25/ICMP/icmp-claude-3-5-sonnet-20241022.rs
use nom::{
    bytes::complete::take,
    combinator::map,
    number::complete::{be_u16, be_u32, be_u8},
    IResult,
};
use std::env;
use std::fs::File;
use std::io::Read;

#[derive(Debug)]
pub enum IcmpType {
    EchoReply,
    DestinationUnreachable,
    SourceQuench,
    Redirect,
    EchoRequest,
    TimeExceeded,
    ParameterProblem,
    Timestamp,
    TimestampReply,
    InformationRequest,
    InformationReply,
    Unknown(u8),
}

#[derive(Debug)]
pub enum IcmpCode {
    Zero,
    NetUnreachable,
    HostUnreachable,
    ProtocolUnreachable,
    PortUnreachable,
    FragmentationNeeded,
    SourceRouteFailed,
    RedirectNetwork,
    RedirectHost,
    RedirectTosNetwork,
    RedirectTosHost,
    TtlExceeded,
    FragmentReassemblyTimeExceeded,
    Unknown(u8),
}

#[derive(Debug)]
pub enum RestOfHeader {
    EchoData { identifier: u16, sequence: u16 },
    ParameterProblem { pointer: u8, unused: [u8; 3] },
    RedirectGateway { gateway_addr: u32 },
    TimestampData { identifier: u16, sequence: u16 },
    None,
}

#[derive(Debug)]
pub struct IcmpPacket {
    typ: IcmpType,
    code: IcmpCode,
    checksum: u16,
    rest_of_header: RestOfHeader,
    data: Vec<u8>,
}

fn parse_icmp_type(input: &[u8]) -> IResult<&[u8], IcmpType> {
    map(be_u8, |b| match b {
        0 => IcmpType::EchoReply,
        3 => IcmpType::DestinationUnreachable,
        4 => IcmpType::SourceQuench,
        5 => IcmpType::Redirect,
        8 => IcmpType::EchoRequest,
        11 => IcmpType::TimeExceeded,
        12 => IcmpType::ParameterProblem,
        13 => IcmpType::Timestamp,
        14 => IcmpType::TimestampReply,
        15 => IcmpType::InformationRequest,
        16 => IcmpType::InformationReply,
        n => IcmpType::Unknown(n),
    })(input)
}

fn parse_icmp_code<'a>(typ: &IcmpType, input: &'a [u8]) -> IResult<&'a [u8], IcmpCode> {
    map(be_u8, |code| match (typ, code) {
        (IcmpType::DestinationUnreachable, 0) => IcmpCode::NetUnreachable,
        (IcmpType::DestinationUnreachable, 1) => IcmpCode::HostUnreachable,
        (IcmpType::DestinationUnreachable, 2) => IcmpCode::ProtocolUnreachable,
        (IcmpType::DestinationUnreachable, 3) => IcmpCode::PortUnreachable,
        (IcmpType::DestinationUnreachable, 4) => IcmpCode::FragmentationNeeded,
        (IcmpType::DestinationUnreachable, 5) => IcmpCode::SourceRouteFailed,
        (IcmpType::Redirect, 0) => IcmpCode::RedirectNetwork,
        (IcmpType::Redirect, 1) => IcmpCode::RedirectHost,
        (IcmpType::Redirect, 2) => IcmpCode::RedirectTosNetwork,
        (IcmpType::Redirect, 3) => IcmpCode::RedirectTosHost,
        (IcmpType::TimeExceeded, 0) => IcmpCode::TtlExceeded,
        (IcmpType::TimeExceeded, 1) => IcmpCode::FragmentReassemblyTimeExceeded,
        (_, 0) => IcmpCode::Zero,
        (_, n) => IcmpCode::Unknown(n),
    })(input)
}

fn parse_rest_of_header<'a>(typ: &IcmpType, input: &'a [u8]) -> IResult<&'a [u8], RestOfHeader> {
    match typ {
        IcmpType::EchoReply | IcmpType::EchoRequest => {
            let (input, identifier) = be_u16(input)?;
            let (input, sequence) = be_u16(input)?;
            Ok((input, RestOfHeader::EchoData { identifier, sequence }))
        }
        IcmpType::ParameterProblem => {
            let (input, pointer) = be_u8(input)?;
            let (input, unused) = map(take(3usize), |b: &[u8]| [b[0], b[1], b[2]])(input)?;
            Ok((input, RestOfHeader::ParameterProblem { pointer, unused }))
        }
        IcmpType::Redirect => {
            let (input, gateway_addr) = be_u32(input)?;
            Ok((input, RestOfHeader::RedirectGateway { gateway_addr }))
        }
        IcmpType::Timestamp | IcmpType::TimestampReply => {
            let (input, identifier) = be_u16(input)?;
            let (input, sequence) = be_u16(input)?;
            Ok((input, RestOfHeader::TimestampData { identifier, sequence }))
        }
        _ => Ok((input, RestOfHeader::None)),
    }
}

fn parse_icmp_packet(input: &[u8]) -> IResult<&[u8], IcmpPacket> {
    let (input, typ) = parse_icmp_type(input)?;
    let (input, code) = parse_icmp_code(&typ, input)?;
    let (input, checksum) = be_u16(input)?;
    let (input, rest_of_header) = parse_rest_of_header(&typ, input)?;
    let (input, data) = map(take(input.len()), |b: &[u8]| b.to_vec())(input)?;

    Ok((
        input,
        IcmpPacket {
            typ,
            code,
            checksum,
            rest_of_header,
            data,
        },
    ))
}

fn main() -> std::io::Result<()> {
    let args: Vec<String> = env::args().collect();
    if args.len() != 2 {
        eprintln!("Usage: {} <icmp_binary_file>", args[0]);
        std::process::exit(1);
    }

    let mut file = File::open(&args[1])?;
    let mut buffer = Vec::new();
    file.read_to_end(&mut buffer)?;

    match parse_icmp_packet(&buffer) {
        Ok((remaining, packet)) => {
            println!("Parsed ICMP packet: {:?}", packet);
            println!("Remaining bytes: {} bytes", remaining.len());
        }
        Err(e) => eprintln!("Failed to parse ICMP packet: {:?}", e),
    }

    Ok(())
}